src/controllers/cart.controller.js

```diff
@@ -0,0 +1,137 @@
+  const { productId, quantity } = req.body;
+  let { userId } = req;
+  try {
+    await doesProductExist(productId);
+    let cart = await findCartByUserId(userId);
+    if (cart) {
+      cart = await handleDuplicateCartItems(cart, productId, quantity);
+      if (quantity < 1) {
+        return res.json({
+          status: "SUCCESS",
+          data: cart,
+          message: `Product ${productId} removed from cart successfully`,
+        });
+      }
+      return res.json({
+        status: "SUCCESS",
+        data: cart,
+        message: `Product ${productId} of quantity ${quantity} added to cart successfully`,
+      });
+    }
+    const newCart = await createNewCart(userId, productId, quantity);
+    return res.json({
+      status: "SUCCESS",
+      data: newCart,
+      message: `New cart created with product ${productId} of quantity ${quantity} successfully`,
+    });
+  } catch (err) {
+    next(err);
+  }
+};
+
+const updateItemQuantity = async (req, res, next) => {
+  const { productId, quantity } = req.body;
+  const { userId } = req;
+  let cart = await findCartByUserId(userId);
+  try {
+    await doesProductExist(productId);
+    if (cart) {
+      cart = await handleUpdateCartItem(cart, productId, quantity);
+      if (quantity < 1) {
+        return res.json({
+          status: "SUCCESS",
+          data: cart,
+          message: `Product ${productId} removed from cart successfully`,
+        });
+      }
+      return res.json({
+        status: "SUCCESS",
+        data: cart,
+        message: `Product ${productId} of quantity ${quantity} updated successfully`,
+      });
+    }
+    const newCart = await createNewCart(userId, productId, quantity);
+    return res.json({
+      status: "SUCCESS",
+      data: newCart,
+      message: `New cart created with product ${productId} of quantity ${quantity} successfully`,
+    });
+  } catch (err) {
+    next(err);
+  }
+};
+
+const updateCart = async (req, res, next) => {
+  const cartItems = req.body;
+  const { userId } = req;
+  try {
+    if (!(cartItems instanceof Array)) {
+      throw new HttpError(400, "cart should be an array");
+    }
+    let cart = await findCartByUserId(userId);
+    if (cart) {
+      for (let i = 0; i < cartItems.length; i++) {
+        let item = cartItems[i];
+        await doesProductExist(item.product._id);
+        cart = await handleUpdateCartItem(
+          cart,
+          item.product._id,
+          item.quantity
+        );
+      }
+      return res.json({
+        status: "SUCCESS",
+        data: cart,
+        message: "Cart updated successfully",
+      });
+    }
+    let newCart = new Cart({ user: userId });
+    cartItems.forEach((item) => {
+      newCart.items.push({
+        product: item.product._id,
+        quantity: item.quantity,
+      });
+    });
+    await newCart.populate("items.product").execPopulate();
+    newCart.calculateTotalPrice();
+    newCart = await newCart.save();
+    console.log({ newCart });
```

Comment:
Create a Logger utility function for logging. May use chalk for different purposes (INFO, SERVER_ERROR, CLIENT_ERROR etc).

Reply:
Resolved.